superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+- A minimal implementation for the purpose of fault-proving the message dependencies.
+
+### Workload approaches
+
+In the case of low-resource users, and increasingly so with a growing number of interoperating chains,
+distributing the verification should be supported.
+This can be distributed while preserving a high level of security with the help of light-client protocols
+that serve the same message-verification interface.
+
+A common data-availability layer between interoperating chains allows for the light-clients to be computation-only.
+
+As an individual verifier, i.e. a subjective view of the L2 chain,
+it is possible to verify any form of offchain computation proof of choice.
+
+As an L1 smart-contract verifier, i.e. an objective view  of the chain (insofar the bridge is considered canonical),
+it is possible to use any combination of onchain computation proofs,
+without relying on secondary security assumptions for the L2 inputs.
+
+For users that do run multiple nodes and/or nodes of multiple L2 chains,
+the interface can be used to de-duplicate verification work.
+Since the L2 verification is inter-dependent, the verification work largely overlaps,
+and is thus better served by a shared service.
+
+The Superchain Backend Interface abstracts all of the above away,
+such that the right form of verification can be plugged-in in each context.
+
+### Data
+
+What are the constraints for the backend service implementation? How many receipts do we expect?
+
+Currently OP Mainnet has 120M blocks, or about ~4.5 GB of 32 byte block hashes and indexing structures.
+
+1 hour of block hashes amounts to:
+- single L2 chain: `60*60/2 * 32 = 58 KB`
+- 100 L2 chains = 5.8 MB
+
+There is an order of magnitude more log-events however:
+- Worst case @ 375 gas / log:
+  - 30M gas (current limit) -> ~80K log events / block.
+  - 5M gas (current OP-Mainnet target throughput) -> ~13K log events / block.
+- OP-Mainnet today:
+  - OP Mainnet logs per day: [https://dune.com/queries/3699891/6225381/](https://dune.com/queries/3699891/6225381/)
+    OP Mainnet at 4M logs per day. -> About 93 logs per block.
+- Other L2 chains:
+  - Base at 11M logs per day.
+  - Zora at 0.5-2M logs per day.
+  - Arbitrum at 8M logs per day.
+  - Ethereum L1 at 3M logs per day.
+
+The history of these log events sums up to:
+- OP Mainnet history (from Dune query) = 1.9B logs.
+- Size of just 32 bytes per log = ~ 60 GB.
+- Size of receipts in geth (from recent `db inspect`) = 80+ GB.
+
+Caching a day worth of event log hashes is ~130 MB for OP Mainnet.
+The worst-case, if a chain is attacked with unbalanced log-event usage, has significantly more logs than that.
+This would amount to `13k / 93 = ~ 140` times as many log-events,
+or about 18 GB of log-event hashes for a full-day full-capacity attack on one standard-throughput chain.
+
+A standalone-service that prefetches the logs should be able to support the above:
```

Comment:
How will this service scale? Will it support some kind of single-writer multiple-reader pattern to allow us to use multiple k8s replicas, or will we need to load-balance multiple distinct instances of this service (and therefore have some "consensus-awareness" concept in the load balancer)? Alternatively, could it support some kind of shared database?

We don't have to build for scale now, but as we've seen with the recent blob archiver outages we will need some degree of redundancy before going to prod. This service is a mission-critical component of  each interoperable chain, and we need to ensure that it stays up.

Reply:
I want to add that prefetching initiating logs might make sense in a world where users are unpredictably pulling all sorts of events as executing messages.

Maybe this is the case but my hunch is that at least initially we probably will see an extremely small subset of origin addresses that are pulled in as executing messages (bridge & cdm contracts).

Reply:
Added to the notes